flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java

```diff
@@ -0,0 +1,263 @@
+			// This ensures strict sequential processing of requests.
+			// If we send new requests immediately we can no longer make assumptions about the order in which responses
+			// arrive, due to which the handler cannot know which future he should complete (not to mention what response
+			// type to read).
+			CompletableFuture<P> nextFuture = lastFuture
+				.handleAsync((f, e) -> submitRequest(httpRequest, messageHeaders), directExecutor)
+				.thenCompose((future) -> future);
+
+			lastFuture = nextFuture;
+			return nextFuture;
+		}
+	}
+
+	private <M extends MessageHeaders<R, P, U>, U extends ParameterMapper, R extends RequestBody, P extends ResponseBody> CompletableFuture<P> submitRequest(FullHttpRequest httpRequest, M messageHeaders) {
+		CompletableFuture<P> responseFuture = handler.expectResponse(messageHeaders.getResponseClass());
+
+		try {
+			// write request
+			Channel channel = bootstrap.connect(configuredTargetAddress, configuredTargetPort).sync().channel();
+			channel.writeAndFlush(httpRequest);
+			channel.closeFuture();
+		} catch (InterruptedException e) {
+			return FutureUtils.completedExceptionally(e);
+		}
+		return responseFuture;
+	}
+
+	@ChannelHandler.Sharable
+	private static class ClientHandler extends SimpleChannelInboundHandler<Object> {
+
+		private volatile ExpectedResponse<? extends ResponseBody> expectedResponse = null;
+
+		@Override
+		protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
+			if (msg instanceof FullHttpResponse) {
+				completeFuture((FullHttpResponse) msg);
+			} else {
+				LOG.error("Implementation error: Received a response that wasn't a FullHttpResponse.");
+			}
+		}
+
+		private <J extends ResponseBody> void completeFuture(FullHttpResponse msg) throws IOException {
+			ByteBuf content = msg.content();
+
+			JsonNode rawResponse;
+			try {
+				InputStream in = new ByteBufInputStream(content);
+				rawResponse = objectMapper.readTree(in);
+				LOG.debug("Received response {}.", rawResponse);
+			} catch (JsonMappingException | JsonParseException je) {
+				LOG.error("Response was not valid JSON.", je);
+				return;
+			}
+
+			final ExpectedResponse<J> expectedResponse = (ExpectedResponse<J>) this.expectedResponse;
+			this.expectedResponse = null;
+
+			if (expectedResponse == null) {
+				LOG.error("Received response, but we didn't expect any. Response={}", rawResponse);
+				return;
+			}
+
+			try {
+				try {
+					J response = objectMapper.treeToValue(rawResponse, expectedResponse.responseClass);
+					expectedResponse.responseFuture.complete(response);
+				} catch (JsonProcessingException jpe) {
+					// the received response did not matched the expected response type
+
+					// lets see if it is an ErrorResponse instead
+					try {
+						ErrorResponseBody error = objectMapper.treeToValue(rawResponse, ErrorResponseBody.class);
+						expectedResponse.responseFuture.completeExceptionally(new RuntimeException(error.errors.toString()));
```

Comment:
Let's not use `RuntimeExceptions`. I think it's better to introduce a `RestClientException` which extends the `FlinkException`.